### ISSUE TITLE
Oauth `state` optional param, but seems to be required by Meteor

### ISSUE THREAD

**Issue by adamgins** [meteor/meteor#8678]
I am getting the following issue when trying to use OAuth to have IBM Connections Authenticate a Meteor user.

I have created a package to do this, modelled off @robfallows http://robfallows.github.io/2015/12/17/writing-an-oauth-2-handler.html 

I am getting the following error:

```
W20170508-19:57:49.550(10) (oauth.js:101) Unable to base64 decode state from OAuth query: undefined
W20170508-19:57:49.554(10) (oauth.js:101) Unable to base64 decode state from OAuth query: undefined
W20170508-19:57:49.556(10) (oauth.js:431) Error in OAuth Server: Unexpected token a
```

IBM Connection does not support the `state` param. 

I am using Meteor 1.4.4.1

My client code looks like:

```
/**
 * Created by adamginsburg on 3/04/2015.
 */
IBMConnectionsCloudOauth = {};
IBMConnectionsCloudOauth.requestCredential = function (options, credentialRequestCompleteCallback) {
    debugger;
    // support both (options, callback) and (callback).
    console.log("IBMConnectionsCloudOauth", options, "credentialRequestCompleteCallback:", credentialRequestCompleteCallback);

    if (!credentialRequestCompleteCallback && typeof options === 'function') {

        credentialRequestCompleteCallback = options;
        options = {};
    }

    var config = ServiceConfiguration.configurations.findOne({service: 'ibm-connections-cloud'});
    if (!config) {

        credentialRequestCompleteCallback && credentialRequestCompleteCallback(
            new ServiceConfiguration.ConfigError());
        return;
    }

    var credentialToken = Random.secret();
    var mobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|Windows Phone/i.test(navigator.userAgent);
    var display = mobile ? 'touch' : 'popup';

    var scope = "email";
    if (options && options.requestPermissions)
        scope = options.requestPermissions.join(',');
    console.log("IBMConnectionsCloudOauth 2 config:", config);
    var loginStyle = OAuth._loginStyle('ibm-connections-cloud', config, options);

    var loginUrl =
        'https://apps.ce.collabserv.com/manage/oauth2/authorize?client_id=' + config.clientId +
        '&callback_uri=' + Meteor.absoluteUrl("_oauth/ibm-connections-cloud") +
        '&response_type=' + "code"
        //'&state=' + OAuth._stateParam(loginStyle, credentialToken);


    OAuth.launchLogin({
        loginService: "ibm-connections-cloud",
        loginStyle: loginStyle,
        loginUrl: loginUrl,
        credentialRequestCompleteCallback: credentialRequestCompleteCallback,
        credentialToken: credentialToken
    });
};


Meteor.loginWithIBMConnectionsCloud = function (options, callback) {
    // support a callback without options
    if (!callback && typeof options === "function") {
        callback = options;
        options = null;
    }


    var credentialRequestCompleteCallback = Accounts.oauth.credentialRequestCompleteHandler(callback);
    console.log("inside Meteor.loginWithIBMConnectionsCloud", credentialRequestCompleteCallback);
    IBMConnectionsCloudOauth.requestCredential(options, credentialRequestCompleteCallback);
};

```

The server code (work in progress):

```
IBMConnectionsCloudOauth = {};

IBMConnectionsCloudOauth.whitelistedFields = ['user_id', 'name', 'first_name', 'last_name', 'email'];

OAuth.registerService('ibm-connections-cloud', 2, null, function (oauthBinding) {
    //console.log("OAuth.registerService 2:", oauthBinding)
    var response = getTokenResponse(oauthBinding);
    var accessToken = response.accessToken;
    var identity = getIdentity(accessToken);
    console.log("identity", identity);

    var serviceData = {
        id: identity.user_id,
        //screenName: identity.screen_name,
        accessToken: OAuth.sealSecret(oauthBinding.accessToken),
        accessTokenSecret: OAuth.sealSecret(oauthBinding.accessTokenSecret)
    };

    // include helpful fields from twitter
    var fields = _.pick(identity, IBMConnectionsCloudOauth.whitelistedFields);
    _.extend(serviceData, fields);
    return {
        serviceData: serviceData,
        options: {
            profile: {
                name: identity.name,

            }
        }
    };
});


IBMConnectionsCloudOauth.retrieveCredential = function (credentialToken, credentialSecret) {
    return OAuth.retrieveCredential(credentialToken, credentialSecret);
};

// returns an object containing:
// - accessToken
// - expiresIn: lifetime of token in seconds
var getTokenResponse = function (query) {

    debugger;
    var config = ServiceConfiguration.configurations.findOne({service: 'ibm-connections-cloud'});
    if (!config) {
        throw new ServiceConfiguration.ConfigError("Service not configured");
    }

    var responseContent;
    console.log("getTokenResponse query", query);

    try {
        // Request an access token
        responseContent = HTTP.post(
            "https://apps.ce.collabserv.com/manage/oauth2/token", {
                params: {
                    client_id: config.clientId,
                    client_secret:  OAuth.openSecret(config.secret),
                    code: query.code,
                    grant_type: 'authorization_code',
                    callback_uri: Meteor.absoluteUrl("_oauth/ibm-connections-cloud"),
                    //state: query.state
                }
            }).content;
        console.log("responseContent:", responseContent);


    } catch (err) {
        throw _.extend(new Error("Failed to complete OAuth handshake with ibm-connections-cloud. " + err.message),
            {response: err.response});
    }
    debugger;
    var parsedResponse = JSON.parse(responseContent);

    if (!parsedResponse.access_token) {
        throw new Error("Failed to complete OAuth handshake with ibm-connections-cloud " +
        "-- can't find access token in HTTP response. " + parsedResponse);
    }
    return {
        accessToken: parsedResponse.access_token,
        expires_in: (+new Date) + (1000 * parsedResponse.expires_in)
    }


};

var getIdentity = function (accessToken) {


    try {
        console.log("about to try get user info");
        var responseData = HTTP.get(
            "https://apps.ce.collabserv.com/manage/oauth/getUserIdentity", {
                headers: {Authorization: 'Bearer ' + accessToken}
            }).data;
        console.log("***************** responseContent 2:", responseData);


        var microsoftAccessToken = accessToken;
        var microsoft_id = responseData.id;


    } catch (err) {
        throw _.extend(new Error("Failed to get user details. " + err.message),
            {response: err.response});
    }


    return {
        accessToken: microsoftAccessToken,
        user_id: microsoft_id,
        name: responseData.name,
        email: responseData.emails.preferred

    };
}

```

Package.js excerpt:

```
Package.onUse(function(api) {
  api.use('http', ['client', 'server']);
  api.use('templating', 'client');
  api.use('oauth1', ['client', 'server']);
  api.use('oauth2', ['client', 'server']);
  api.use('oauth', ['client', 'server']);
  api.use('random', 'client');
  api.use('underscore', 'server');
  api.use('service-configuration', ['client', 'server']);

  api.export('IBMConnectionsCloudOauth');
  api.addFiles('ibm_connections_cloud_client.js','client');
  api.addFiles('ibm_connections_cloud_server.js','server');


});
```

versions:

```
accounts-base@1.2.16
accounts-facebook@1.1.1
accounts-google@1.1.2
accounts-oauth@1.1.15
accounts-password@1.3.5
accounts-twitter@1.2.1
adamgins:microsoft-oauth@0.0.2
alisalaah:hogan@1.0.1
allow-deny@1.0.5
autoupdate@1.3.12
babel-compiler@6.18.1
babel-runtime@1.0.1
barbatus:stars-rating@1.1.1
base64@1.0.10
binary-heap@1.0.10
blaze@2.3.2
blaze-html-templates@1.1.2
blaze-tools@1.0.10
boilerplate-generator@1.0.11
bozhao:accounts-instagram@0.2.2
browser-policy@1.1.0
browser-policy-common@1.0.11
browser-policy-content@1.1.0
browser-policy-framing@1.1.0
buzzy-buzz:avatar@0.9.2
buzzy-buzz:common-globals@0.1.3
buzzy-buzz:google-api@1.0.3
buzzy-buzz:ibm-connections-cloud-oauth@0.0.1
buzzy-buzz:resources-core@0.1.6
caching-compiler@1.1.9
caching-html-compiler@1.1.2
callback-hook@1.0.10
check@1.2.5
coffeescript@1.12.3_1
copleykj:jquery-autosize@1.17.8
cosmos:browserify@0.10.0
crosswalk@1.7.1
dburles:google-maps@1.1.5
ddp@1.2.5
ddp-client@1.3.4
ddp-common@1.2.8
ddp-rate-limiter@1.0.7
ddp-server@1.3.14
deps@1.0.12
diff-sequence@1.0.7
dispatch:deep-link@4.0.4
djedi:sanitize-html@1.11.3
ecmascript@0.7.2
ecmascript-runtime@0.3.15
edgee:slingshot@0.7.1
ejson@1.0.13
email@1.2.1
es5-shim@4.6.15
facebook-config-ui@1.0.0
facebook-oauth@1.3.0
fastclick@1.0.13
force-ssl@1.0.14
force-ssl-common@1.0.14
fortawesome:fontawesome@4.7.0
froala:editor@2.5.1
froala:editor-reactive@2.5.1
froatsnook:request@2.74.0
geojson-utils@1.0.10
google-config-ui@1.0.0
google-oauth@1.2.1
handlebars@1.0.7
harrison:papa-parse@1.1.7
hot-code-push@1.0.4
html-tools@1.0.11
htmljs@1.0.11
http@1.2.12
huttonr:bootstrap3@3.3.6_13
huttonr:bootstrap3-assets@3.3.6_3
id-map@1.0.9
iron:controller@1.0.12
iron:core@1.0.11
iron:dynamic-template@1.0.12
iron:layout@1.0.12
iron:location@1.0.11
iron:middleware-stack@1.1.0
iron:router@1.1.2
iron:url@1.1.0
jamielob:reloader@1.2.3
jeremy:selectize@0.12.1_5
johnantoni:meteor-appear@0.0.2
jparker:crypto-core@0.1.0
jparker:crypto-hmac@0.1.0
jparker:crypto-md5@0.1.1
jparker:crypto-sha256@0.1.1
jparker:gravatar@0.5.1
jquery@1.11.10
launch-screen@1.1.1
less@2.7.9
linto:jquery-ui@1.11.2
livedata@1.0.18
localstorage@1.0.12
logging@1.1.17
lookback:emails@0.7.7
lookback:html-to-text@2.1.3_2
lukemadera:social-share@1.2.0
manuelschoebel:cheerio@0.0.1
maxharris9:connection-banner@0.5.0
mdg:geolocation@1.3.0
meteor@1.6.1
meteor-base@1.0.4
meteorhacks:aggregate@1.3.0
meteorhacks:async@1.0.0
meteorhacks:collection-utils@1.2.0
meteorhacks:inject-initial@1.0.4
meteorhacks:meteorx@1.4.1
meteorhacks:npm@1.5.0
meteorhacks:picker@1.0.3
meteorhacks:ssr@2.2.0
meteorhacks:subs-manager@1.6.4
meteorhacks:unblock@1.1.0
minifier-css@1.2.16
minifier-js@2.0.0
minimongo@1.1.0-ddp-batching-beta.0
mizzao:timesync@0.5.0
mobile-experience@1.0.4
mobile-status-bar@1.0.14
modules@0.8.1
modules-runtime@0.7.10
momentjs:moment@2.18.1
mongo@1.1.16
mongo-id@1.0.6
mongo-livedata@1.0.12
mquandalle:ismobile@0.1.0
mrt:q@1.0.1
mrt:twilio-meteor@1.1.0
mrt:twit@0.2.0
natestrauser:filepicker-plus@2.0.0
npm-bcrypt@0.9.2
npm-container@1.2.0
npm-mongo@2.2.24
oauth@1.1.13
oauth1@1.1.11
oauth2@1.1.11
observe-sequence@1.0.16
okgrow:analytics@2.1.3
okgrow:router-autoscroll@0.1.8
ordered-dict@1.0.9
peerlibrary:aws-sdk@2.4.9_1
peerlibrary:blocking@0.5.2
peppelg:bootstrap-3-modal@1.0.4
pfafman:filesaver@1.3.2
practicalmeteor:chai@2.1.0_1
practicalmeteor:loglevel@1.2.0_2
practicalmeteor:mocha@2.4.5_6
practicalmeteor:mocha-core@1.0.1
practicalmeteor:sinon@1.14.1_2
promise@0.8.8
raix:eventemitter@0.1.3
raix:eventstate@0.0.4
raix:handlebar-helpers@0.2.5
raix:push@3.1.0-pre.1
raix:ui-dropped-event@0.0.7
random@1.0.10
rate-limit@1.0.8
reactive-dict@1.1.8
reactive-var@1.0.11
reload@1.1.11
retry@1.0.9
reywood:publish-composite@1.5.1
richsilv:owl-carousel@0.1.0
risul:moment-timezone@0.5.7
routepolicy@1.0.12
sacha:juice@0.1.4
sacha:spin@2.3.1
service-configuration@1.0.11
session@1.1.7
sha@1.0.9
shell-server@0.2.3
simple:reactive-method@1.0.2
spacebars@1.0.15
spacebars-compiler@1.1.2
srp@1.0.10
standard-minifier-css@1.3.4
standard-minifier-js@2.0.0
templating@1.3.2
templating-compiler@1.3.2
templating-runtime@1.3.2
templating-tools@1.1.2
tmeasday:test-reporter-helpers@0.2.1
tracker@1.1.2
tsega:bootstrap3-datetimepicker@4.17.37_1
twitter-config-ui@1.0.0
twitter-oauth@1.2.0
ui@1.0.13
underscore@1.0.10
url@1.1.0
webapp@1.3.15
webapp-hashing@1.0.9

```

**Comment by adamgins** [comment]
Hi, i have kind of solved the issue... 

the errors

```
Unable to base64 decode state from OAuth query: undefined
W20170508-19:57:49.554(10) (oauth.js:101) Unable to base64 decode state from OAuth query: undefined
```

we red herrings, the real issue was `responseContent` was not JSON. 

That said, not sure if the way the satte param is checked and err'd is correct, as it's not a required param in the spec... so it kind of threw me off the trail off the real error.

**Comment by abernix** [comment]
Glad you got this fixed!  The `state` parameter is not required, but it's certainly suggested.  More reading on it is [available here](https://auth0.com/docs/protocols/oauth2/oauth-state), and I believe we've implemented it correctly.

Thanks for filing this issue and following up on it!

**Comment by steve-ross** [comment]
I'm having a similar issue with making my own OAuth provider. Getting the following error:
`W20171127-17:09:19.366(-5) (oauth.js:431) Error in OAuth Server: Match error: Expected string, got undefined`

I dumped my return from my provider and none of the fields are undefined (only thing i see that might be an issue is that my ID is a number?):
```
  const returnData = {
    serviceData: {
      id: responseData.user.id,
      accessToken: OAuth.sealSecret(responseData.access_token),
      email: responseData.user.email || '',
      username: responseData.user.username,
      context: responseData.context,
      scope: responseData.scope,
      options: {profile: {name: responseData.user.username}},
      emails: [responseData.user.email],
    },
  };
```

**Comment by steve-ross** [comment]
Also worth noting that my flow is being initiated from a third party app and not the login UI. The 3rd party is initiating by hitting `localhost/_oauth/myendpoint` and does not set the state param. I receive a warning about that (3 times) but, then I get that final error.

**Comment by adamgins** [comment]
I had to do some pre-processing, so I could deal with the anomalies of the oauth provider and then construct things in the way Meteor expects it... BTW, I could not get the Meteor stuff to work without the state param. 

initially I create login token and state and save it locally. 

eg I have a route `/pre-oauth/<some params>` as the call back URL for my provider. I then construct a redirect to the `_oauth` route... eg

```
var urlRedirect = Meteor.absoluteUrl("_oauth/<name of provider>") +
		'?code=' + this.params.query.code +
		'&credentialToken=' + credentialToken +
		'&state=' + connectionsState
```

Hope that helps

**Comment by steve-ross** [comment]
Ah makes sense, thought I'd have to do something like that, guessing this is because clientToken has to be set in order to do OAuth "the meteor way"

**Comment by adamgins** [comment]
not sure if other ways to make it work, but that's the way i got it to work.. ie I append 

```
'&credentialToken=' + credentialToken +
		'&state=' + connectionsState
```

**Comment by steve-ross** [comment]
@adamgins getting further... now I'm getting:
```W20171127-20:00:21.439(-5) (oauth.js:108) Unable to parse state from OAuth query: [object JSON]
W20171127-20:00:21.441(-5) (oauth.js:431) Error in OAuth Server: invalid_credential_token_or_secret```

I wrote the redirect route in express:
```
  app.get('/auth/myoauth', function(req, res) {
    console.log('query is?', req.query);
    const {code, context, scope} = req.query;
    const credentialToken = Random.secret();

    const state = {
      loginStyle: 'redirect',
      credentialToken,
      isCordova: Meteor.isCordova,
      redirectUrl: `${Meteor.absoluteUrl()}auth/myoauth`,
    };
    const encodedState = Buffer.from(JSON.toString(state)).toString('base64');
    const redirectURL = `${Meteor.absoluteUrl('_oauth/bigcommerce')}?code=${
      code
    }&context=${context}&scope=${scope}&state=${encodedState}&credentialToken=${
      credentialToken
    }`;
    res.redirect(redirectURL);
  });
```

**Comment by adamgins** [comment]
ok, it may be were you're creating the credential token.

what I do iin the client requestCredential routing I create both state and credentialToken, save to disk eg `window.localStorage.setItem` 

then I call 

```
	OAuth.launchLogin({
		loginService: "someservice",
		loginStyle: loginStyle,
		loginUrl: loginUrl,
		credentialRequestCompleteCallback: credentialRequestCompleteCallback,
		credentialToken: credentialToken
	});
```

then in the route I retrieve the token I used above. Ie so only the current user can get the credToken on the specific computer and then user it as above https://github.com/meteor/meteor/issues/8678#issuecomment-347357060

so the state and token will match between the `launchLogin` and the `_oauth` call you redirect to.

**Comment by steve-ross** [comment]
I had some bad formatting... please see above (again) Note the json error

**Comment by steve-ross** [comment]
@adamgins I had some bad formatting... please see above (again) Note the json error. I see your point though I need to store the token locally

**Comment by adamgins** [comment]
I am not 100% sure.. but I think you need to start the dance (ie OAuth.launchLogin) with the same token you use later so it can validate it's all OK.

**Comment by steve-ross** [comment]
So here is how I accomplished this in react (BigCommerce OAuth is the provider)
Required meteor packages (this stumped me):

```
accounts-base
service-configuration
accounts-oauth
oauth
```

```javascript
// BigCommercePageContainer.jsx
import {Meteor} from 'meteor/meteor';
import {ServiceConfiguration} from 'meteor/service-configuration';
import {withTracker} from 'meteor/react-meteor-data';

import BigCommercePage from '../pages/BigCommercePage';

const BigCommercePageContainer = withTracker(() => {
  const servicesHandle = Meteor.subscribe('meteor.loginServiceConfiguration');
  const loading = !servicesHandle.ready();
  const serviceConfig = ServiceConfiguration.configurations.findOne({
    service: 'bigcommerce',
  });
  const serviceConfigExists = !loading && !!serviceConfig;
  return {
    loading,
    serviceConfig,
    serviceConfigExists,
  };
})(BigCommercePage);

export default BigCommercePageContainer;
```

```javascript
//BigCommercePage.jsx
import {Meteor} from 'meteor/meteor';
import {Random} from 'meteor/random';
import {OAuth} from 'meteor/oauth';
import React from 'react';
import PropTypes from 'prop-types';
import qs from 'qs';

export default class BigCommercePage extends React.Component {
  // eslint-disable-line react/prefer-stateless-function, max-len
  componentWillReceiveProps(nextProps) {
    const {loading, serviceConfig, serviceConfigExists} = nextProps;
    if (!loading && serviceConfigExists && serviceConfig) {
      // initiate OAuth flow
      const params = qs.parse(this.props.location.search, { // eslint-disable-line 
        ignoreQueryPrefix: true,
      });
      const {code, context, scope} = params;

      if (code && scope) {
        const credentialToken = Random.secret();
        const loginStyle = OAuth._loginStyle('bigcommerce', serviceConfig); // eslint-disable-line 
        const redirectUrl = Meteor.absoluteUrl('bigcommerce/auth');
        const state = OAuth._stateParam( // eslint-disable-line 
          loginStyle,
          credentialToken,
          redirectUrl,
        );
        const loginUrl =
          `${OAuth._redirectUri('bigcommerce', serviceConfig)}` + // eslint-disable-line 
          `?code=${code}` +
          `&scope=${scope}` +
          `&context=${context}` +
          `&credentialToken=${credentialToken}` +
          `&state=${state}`;

        OAuth.launchLogin({
          loginService: 'bigcommerce',
          loginStyle,
          loginUrl,
          credentialToken,
        });
      } else {
        // assume the app is installed but they are logged out
      }
    } else {
      // if we aren't loading we have a config error
    }
  }

  render() {
    const {loading} = this.props;
    return <div>{loading}</div>;
  }
}

BigCommercePage.defaultProps = {
  loading: true,
  serviceConfig: {},
  serviceConfigExists: false,
};

const {bool, object} = PropTypes;

BigCommercePage.propTypes = {
  loading: bool,
  serviceConfig: object, // eslint-disable-line react/forbid-prop-types
  serviceConfigExists: bool,
};
```

**Comment by martintadashi** [comment]
We're having the exact same issue with Meteor 1.7.

The user goes to the provider site which ends up calling us : 

https://localhost:3000/_oauth/clever?code=0afff4444456666666&scope=read%3Aschool_admins%20read%3Astudent_contacts%20read%3Asections%20read%3Aschools%20read%3Ateachers%20read%3Astudents%20read%3Adistricts%20read%3Auser_id%20read%3Adistrict_admins

This is our handler 

`OAuth.registerService('clever', 2, null, query => {

    const accessToken = getAccessToken(query);
    const identity = getIdentity(accessToken);
    identity.accessToken = accessToken;
    const info = getAllInfo(identity);
    return {
        serviceData: {
            id: info.data.id,
            accessToken: OAuth.sealSecret(accessToken),
            email: info.data.email || '',
            username: info.data.teacher_number
        },
        options: {
            profile: {
                firstName: info.data.name.first + " " + info.data.name.middle,
                lastName: info.data.name.last,
                clever: info.data
            }
        }
    };
});
`

And we get the following errors:

W20181010-16:03:01.861(-3) (oauth.js:100) Unable to base64 decode state from OAuth query: undefined
W20181010-16:03:01.862(-3) (oauth.js:100) Unable to base64 decode state from OAuth query: undefined
W20181010-16:03:01.864(-3) (oauth.js:100) Unable to base64 decode state from OAuth query: undefined
W20181010-16:03:01.866(-3) (oauth.js:424) Error in OAuth Server: Match error: Expected string, got undefined

Any ideas much appreciated!

**Comment by derwaldgeist** [comment]
I am seeing the same problem with Facebook OAuth. It also says: "Invalid verification code format", the error number is 100 and there is a fbtrace_id key included in the message.

**Comment by filipenevola** [comment]
Hi @derwaldgeist could you open a new issue with more detail?

I have projects using Facebook OAuth in production then I believe I can help you.
